src/unittest/Dasm.cc

```diff
@@ -0,0 +1,55 @@
+#include "Dasm.hh"
+#include "EmuTime.hh"
+#include "MSXCPUInterface.hh"
+
+#include "catch.hpp"
+
+#include <iostream>
+
+using namespace openmsx;
+
+TEST_CASE("instructionLength, dasm")
+{
+	std::array<uint8_t, 4> opcode = {0, 0, 0, 0};
+	std::string dest;
+
+	bool end = false;
+	while (!end) {
+		auto len = instructionLength(opcode).value_or(0);
+		assert(1 <= len);
+		assert(len <= 4);
+
+		dest.clear();
+		auto dasmLength = dasm(std::span(opcode.data(), len), 0, dest);
+
+		CHECKED_ELSE( len == dasmLength ) {
+			std::stringstream ss;
+			ss << std::hex;
+
+			for (unsigned j = 0; j < opcode.size(); ++j) {
+				ss << std::setw(2) << std::setfill('0') << int(opcode[j]) << ' ';
+			}
+			ss << '(' << dest << ")\n";
+			INFO( "opcode: " << ss.str());
+
+			auto instructionLength = len;
+			CHECK( dasmLength == instructionLength );
+		}
+
+		// goto next instruction
+		unsigned idx = len - 1;
+		while (true) {
+			if (opcode[idx] == 255) {
+				opcode[idx] = 0;
+				if (idx == 0) {
+					end = true;
+					break;
+				}
+				--idx;
+			} else {
+				++opcode[idx];
+				break;
+			}
+		}
+	}
```

Comment:
In my test I also checked that each call to `dasm()` (with different input in `opcode`) returned a different string. I understand that check won't work in your case (when doing "recovery"). But we want to remove this recovery strategy.

I don't know if it's a good idea to check this. But at least it gave me more confidence that the logic in "goto next instruction" is correct.

Ah, and to make this pass I had to change how `ED 63 nn nn` gets disassembled. Currently it's disassembled to `ld (nn),hl` but that collides with `22 nn nn`. To fix that I changed the disassembly of this longer encoding to `ld (nn),hl*`, but feel free to pick a better string. I don't expect to see this string often (or at all) in normal code.